cmd/kube-proxy/app/server.go

```diff
@@ -591,6 +600,95 @@ func newProxyServer(config *kubeproxyconfig.KubeProxyConfiguration, master strin
 	return s, nil
 }
 
+// checkIPConfig confirms that s has proper configuration for its primary IP family.
+func checkIPConfig(s *ProxyServer, dualStackSupported bool) (error, bool) {
+	var errors []error
+	var badFamily netutils.IPFamily
+
+	if s.PrimaryIPFamily == v1.IPv4Protocol {
+		badFamily = netutils.IPv6
+	} else {
+		badFamily = netutils.IPv4
+	}
+
+	var clusterType string
+	if dualStackSupported {
+		clusterType = fmt.Sprintf("%s-primary", s.PrimaryIPFamily)
+	} else {
+		clusterType = fmt.Sprintf("%s-only", s.PrimaryIPFamily)
+	}
+
+	// Historically, we did not check most of the config options, so we cannot
+	// retroactively make IP family mismatches in those options be fatal. When we add
+	// new options to check here, we should make problems with those options be fatal.
+	fatal := false
+
+	if s.Config.ClusterCIDR != "" {
+		clusterCIDRs := strings.Split(s.Config.ClusterCIDR, ",")
+		if badCIDRs(clusterCIDRs, badFamily) {
+			errors = append(errors, fmt.Errorf("cluster is %s but clusterCIDRs contains only IPv%s addresses", clusterType, badFamily))
+		}
+	}
+
+	if badCIDRs(s.Config.NodePortAddresses, badFamily) {
+		errors = append(errors, fmt.Errorf("cluster is %s but nodePortAddresses contains only IPv%s addresses", clusterType, badFamily))
+	}
+
+	if badCIDRs(s.podCIDRs, badFamily) {
+		errors = append(errors, fmt.Errorf("cluster is %s but node.spec.podCIDRs contains only IPv%s addresses", clusterType, badFamily))
+		if s.Config.DetectLocalMode == kubeproxyconfig.LocalModeNodeCIDR {
+			// This has always been a fatal error
+			fatal = true
+		}
+	}
+
+	if netutils.IPFamilyOfString(s.Config.Winkernel.SourceVip) == badFamily {
+		errors = append(errors, fmt.Errorf("cluster is %s but winkernel.sourceVip is IPv%s", clusterType, badFamily))
+	}
+
+	// In some cases, wrong-IP-family is only a problem when the secondary IP family
+	// isn't present at all.
+	if !dualStackSupported {
+		if badCIDRs(s.Config.IPVS.ExcludeCIDRs, badFamily) {
+			errors = append(errors, fmt.Errorf("cluster is %s but ipvs.excludeCIDRs contains only IPv%s addresses", clusterType, badFamily))
+		}
+
+		if badBindAddress(s.Config.HealthzBindAddress, badFamily) {
```

Comment:
Does this handle "0.0.0.0:<port>" and ":<port>" correctly?  It reads as 4 but accepts on 6, too:

```
$ cat main.go 
package main

import (
	"fmt"
	"net"
	"os"
)

func main() {
	l, err := net.Listen("tcp", "0.0.0.0:9376")
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	l.Accept()
	fmt.Println("done")
}

$ go run main.go &
[1] 3125929

$ netcat -6 localhost 9376
done

[1]+  Done                    go run main.go

$ go run main.go &
[1] 3126055

$ netcat -4 localhost 9376
done

[1]+  Done                    go run main.go

```

Reply:
sorry, yes, I addressed that in the update: `badBindAddress` checks `ip.IsUnspecified()` now, and considers that acceptable regardless of family; [the unit tests check that](https://github.com/kubernetes/kubernetes/pull/119003/commits/3070ca58f1ce8be6057824b8202d7cbc85a97d87#diff-11f277845a3fc64cb4b599b0acb80f48288f9ebe249f4bae4726ea58b51a65d5R849).